### ISSUE TITLE
Handling categorical variables on new data

### ISSUE THREAD

**Issue by krolikowskib** [microsoft/FLAML#1101]
Hey,
I have 2 questions regarding how FLAML handles categorical variables on new data, different from the initial training dataset (for example, during inference after model deployment).

1. Does it handle new categories in categorical features (unseen during training)?
2. SKLearn and XGBoost estimators use ordinal encodings of categorical features. But it seems the categorical codes are extracted during inference ([code](https://github.com/microsoft/FLAML/blob/e3ca95bf8a5cc9c86638dab8ef0eb44bc1f3ee6f/flaml/automl/model.py#L1221)). Doesn't it mean that the encodings will be different when running on a different dataset, thus mixing the categories passed to the model?

**Comment by skzhang1** [comment]
Hi, Thanks for your feedback. 
1. **Does it handle new categories in categorical features (unseen during training)?**
No, it doesn't handle new categories. XGBoost library assumes that category mappings are managed by the application, both in the training/testing phase. In FLAML, when new categories come in the test phase, the categories-encoding map will be different from the map in the training phase if we do not do any process. 
2. **Doesn't it mean that the encodings will be different when running on a different dataset, thus mixing the categories passed to the model?** 
Yes, when running on a different dataset with mixing categories, the encoding will change.   [OrdinalEncoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OrdinalEncoder.html) may be a better choice here. Thanks for your suggestion.
Reference:
https://stackoverflow.com/questions/75698242/when-using-categorical-data-in-xgboost-how-do-i-maintain-the-implied-encoding

**Comment by krolikowskib** [comment]
Thanks for your answer, @skzhang1.

I think that it may be misleading for people who want to reuse the model on a different dataset, like in a production setting.
Even if the categories are the same, the current implemention doesn't guarantee they will be encoded in the same way.

Did you consider to make that more explicit in the documentation or provide a way to easily reuse the best selected model without having to worry about categorical variables?

**Comment by skzhang1** [comment]
Thanks for your suggestion! We will make it clear in the doc. @krolikowskib

**Comment by maayanorner** [comment]
> Thanks for your answer, [@skzhang1](https://github.com/skzhang1).
> 
> I think that it may be misleading for people who want to reuse the model on a different dataset, like in a production setting. Even if the categories are the same, the current implemention doesn't guarantee they will be encoded in the same way.
> 
> Did you consider to make that more explicit in the documentation or provide a way to easily reuse the best selected model without having to worry about categorical variables?

Personally, I think it would be nice to print a warning.
It's difficult for me to imagine a use case where it's a desired behavior, even for research it might lead to meaningless results (e.g., when evaluating on some test set). My argument is that this is just failing silently for most use cases.

**Comment by immu4989** [comment]
Picking this up. I reproduced both failure modes raised here, and they're worse than just a missing doc:

**1. Silent code drift.** The current `DataTransformer.transform()` in `flaml/automl/data.py` re-runs `X[cat_columns] = X[cat_columns].astype("category")` on the predict-time DataFrame, which infers categories from whatever values happen to be present. So if fit data has `gender ∈ {F, M}` (F=0, M=1) and predict data has `gender ∈ {M}` only, M is silently re-encoded as 0 — the model gets the wrong feature value and predictions are garbage.

**2. Unseen-category silent encoding.** A new category at predict time gets a fresh integer code that the model never saw at fit time. There's no warning, no error — the prediction is just wrong.

Repro:

```python
import pandas as pd, numpy as np
from flaml.automl.data import DataTransformer
from flaml.automl.task.factory import task_factory

rng = np.random.RandomState(0)
fit_df = pd.DataFrame({'a': rng.randn(100), 'gender': rng.choice(['M','F'], 100)})
t = DataTransformer()
X_fit, _ = t.fit_transform(fit_df.copy(), pd.Series(rng.randn(100)), task_factory('regression', fit_df, None))
print('fit M code:', X_fit['gender'].cat.codes[X_fit['gender']=='M'].iloc[0])  # → 1

X_pred = t.transform(pd.DataFrame({'a': np.zeros(20), 'gender': ['M']*20}))
print('predict M code:', X_pred['gender'].cat.codes.iloc[0])                  # → 0  ← drift
```

Plan, following @skzhang1's commit upthread (*"We will make it clear in the doc"*) but going slightly further since this is a silent correctness bug:

- Save the per-column category list at fit time on the `DataTransformer` instance.
- At transform time, pin the categories to that list (consistent codes for seen values), map unseen values to the existing `'__NAN__'` sentinel slot, and emit a clear `UserWarning` so production users notice.
- Backward-compatible: old pickled `DataTransformer` instances without the saved category list fall back to today's behavior.

Opening a small PR with the fix and a regression test covering both the drift and the unseen-category cases.
